spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java

```diff
@@ -112,13 +149,62 @@ private CloseableIterable<InternalRow> openChangelogScanTask(ChangelogScanTask t
   CloseableIterable<InternalRow> openAddedRowsScanTask(AddedRowsScanTask task) {
     String filePath = task.file().path().toString();
     SparkDeleteFilter deletes = new SparkDeleteFilter(filePath, task.deletes(), counter());
-    return deletes.filter(rows(task, deletes.requiredSchema()));
+    int[] indexes = indexesInRow(deletes.requiredSchema());
+
+    return CloseableIterable.transform(
+        deletes.filter(rows(task, deletes.requiredSchema())),
+        row -> {
+          InternalRow expectedRow = new GenericInternalRow(columns.length);
+
+          for (int i = 0; i < columns.length; i++) {
+            expectedRow.update(i, row.get(indexes[i], sparkColumnTypes[i]));
+          }
+
+          return expectedRow;
+        });
```

Comment:
When `DeleteFilter::filter` is called, if there are deletes to be applied, the `requiredSchema` could have additional columns beyond the requested schema (e.g., `_pos` is added if there are positional deletes to be applied). Thus the `InternalRow` could have more columns than the requested schema, and we need to exclude the additional columns and only keep the ones for the requested schema.

Reply:
yeah I realized this as well when reading from RowDataReader, looks like if there's position delete present it will always add `pos` metadata column as required in https://github.com/apache/iceberg/blob/684f7a767c2c216a402b60b73d2d55ef605921a0/data/src/main/java/org/apache/iceberg/data/DeleteFilter.java#L260-L263, as well as equality delete id if equality delete are present.

Reply:
Speak of which, i saw this update to `InternalRow` has been repeated here and below for deletedDataFileScanTask, do you know if we can put it into a method instead?

Reply:
The code looks the same, but it is not the same transform as it is a closure, and `indexes` in the each closure is different, as `indexes` is a function of the `SparkDeleteFilter` in each case and that is different.
I didn't try too hard to avoid the code repetition here, as I didn't feel it was worth the effort.

Reply:
I'm adding a method `projectRow`, so that the transform can be written as `row -> projectRow(indexes)`.